libafl_nyx/src/helper.rs

```diff
@@ -0,0 +1,129 @@
+/// [`NyxHelper`] is used to wrap `NyxProcess`
+use std::{
+    fmt::{self, Debug},
+    path::Path,
+    time::Duration,
+};
+
+use libafl::Error;
+use libnyx::{NyxProcess, NyxReturnValue};
+
+pub struct NyxHelper {
+    pub nyx_process: NyxProcess,
+    /// real size of trace_bits
+    pub real_map_size: usize,
+    // real size of the trace_bits
+    pub map_size: usize,
+    /// shared memory with instruction bitmaps
+    pub trace_bits: *mut u8,
+}
+
+const MAX_FILE: u32 = 1024 * 1024;
+#[derive(Clone, Copy, Debug)]
+pub enum NyxProcessType {
+    /// stand alone mode
+    ALONE,
+    /// parallel mode's parent, used to create snapshot
+    PARENT,
+    /// parallel mode's child, consume snapshot and execute
+    CHILD,
+}
+impl NyxHelper {
+    /// create `NyxProcess` and do basic settings
+    /// It will convert instance to parent or child using `parent_cpu_id` when set`parallel_mode`
+    pub fn new(
+        target_dir: &Path,
+        cpu_id: u32,
+        snap_mode: bool,
+        parallel_mode: bool,
+        parent_cpu_id: Option<u32>,
+    ) -> Result<Self, Error> {
+        let sharedir = match target_dir.to_str() {
+            Some(x) => x,
+            None => return Err(Error::illegal_argument("can't convert sharedir to str")),
+        };
+        let work_dir = target_dir.join("workdir");
+        let work_dir = work_dir.to_str().expect("unable to convert workdir to str");
+        let nyx_type = if parallel_mode {
+            let parent_cpu_id = match parent_cpu_id {
+                None => {
+                    return Err(Error::illegal_argument(
+                        "please set parent_cpu_id in nyx parallel mode",
+                    ))
+                }
+                Some(x) => x,
+            };
+            if cpu_id == parent_cpu_id {
+                NyxProcessType::PARENT
+            } else {
+                NyxProcessType::CHILD
+            }
+        } else {
+            NyxProcessType::ALONE
+        };
+
+        let nyx_process = match nyx_type {
+            NyxProcessType::ALONE => NyxProcess::new(sharedir, work_dir, cpu_id, MAX_FILE, true),
+            NyxProcessType::PARENT => {
+                NyxProcess::new_parent(sharedir, work_dir, cpu_id, MAX_FILE, true)
+            }
+            NyxProcessType::CHILD => NyxProcess::new_child(sharedir, work_dir, cpu_id, cpu_id),
+        };
+
+        let mut nyx_process =
+            nyx_process.map_err(|msg: String| -> Error { Error::illegal_argument(msg) })?;
+
+        let real_map_size = nyx_process.bitmap_buffer_size();
+        let map_size = ((real_map_size + 63) >> 6) << 6;
+        let trace_bits = nyx_process.bitmap_buffer_mut().as_mut_ptr();
+        nyx_process.option_set_reload_mode(snap_mode);
+        nyx_process.option_apply();
+
+        // default timeout for initial dry-run
```

Comment:
Is this always the correct time, or should we expose a way to set this?

Reply:
(Specifically, what about a slow target that takes long to start)

Reply:
Change from `2 sec` to `5 sec`
it happens in `new`, so the only way to set it is to add additional param...

Reply:
Maybe rename this function to `with_initial_timeout(..)` and add the param, then add a `new` that just calls down to this fn with default timeout (and specify the default timeout as a `const`)?

Reply:
OK, added